sql/src/main/java/io/crate/planner/consumer/ManyTableConsumer.java

```diff
@@ -89,99 +90,195 @@ public Plan consume(AnalyzedRelation relation, ConsumerContext context) {
+        List<Set<QualifiedName>> sets = new ArrayList<>(explicitJoinedRelations);
+        sets.addAll(implicitJoinedRelations);
+
+        // If no `ORDER BY` present we have no preSort to follow so we return the relations in ordering
+        // obtained by the join conditions (explicit and/or implicit) between them.
+        if (preSorted.isEmpty()) {
+            ObjectIntHashMap<QualifiedName> occurrences =
+                getOccurrencesInJoinConditions(relations, explicitJoinedRelations, implicitJoinedRelations);
+
+            Set<QualifiedName> firstJoinPair = getFirstJoinPair(occurrences, sets);
+            assert firstJoinPair != null : "firstJoinPair should not be null";
+            bestOrder.addAll(firstJoinPair);
+        } else {
+            bestOrder.addAll(preSorted);
+        }
 
-                JoinPair joinPair = JoinPairs.ofRelations(a, b, joinPairs, false);
-                if (joinPair == null) {
-                    // relations are not directly joined, lets check if they are part of an outer join
-                    if (outerJoinRelations.contains(a) || outerJoinRelations.contains(b)) {
-                        // part of an outer join, don't change pairs, permutation not possible
-                        continue outerloop;
-                    } else {
-                        pair.clear();
-                        pair.add(a);
-                        pair.add(b);
-                        joinPushDowns += implicitJoinedRelations.contains(pair) ? 1 : 0;
-                        currentPermutationJoinPairs.add(JoinPair.crossJoin(a, b));
-                    }
-                } else {
-                    if (JoinPairs.joinConditionIncludesRelations(currentPermutationJoinPairs, joinPair)) {
-                        joinPushDowns += 1;
-                    }
-                    currentPermutationJoinPairs.add(JoinPair.crossJoin(a, b));
+        buildBestOrderByJoinConditions(sets, bestOrder);
+        bestOrder.addAll(relations);
+        return bestOrder;
+    }
+
+    private static ObjectIntHashMap<QualifiedName> getOccurrencesInJoinConditions(
+        Collection<QualifiedName> relations,
+        Set<? extends Set<QualifiedName>> explicitJoinedRelations,
+        Set<? extends Set<QualifiedName>> implicitJoinedRelations) {
+
+        List<QualifiedName> qualifiedNames = new ArrayList<>(explicitJoinedRelations.size() + implicitJoinedRelations.size());
+        explicitJoinedRelations.forEach(o -> o.forEach(qualifiedNames::add));
+        implicitJoinedRelations.forEach(o -> o.forEach(qualifiedNames::add));
+
+        ObjectIntHashMap<QualifiedName> occurrences = new ObjectIntHashMap<>(relations.size());
+        for (QualifiedName name : qualifiedNames) {
+            occurrences.putOrAdd(name, 1, 1);
+        }
+        return occurrences;
+    }
+
+    @VisibleForTesting
+    static Set<QualifiedName> getFirstJoinPair(ObjectIntHashMap<QualifiedName> occurrences,
```

Comment:
Seems a bit odd from an API perspective - as it looks like a pure function but also mutates the sets collection.